system/lib/wasmfs/syscalls.cpp

```diff
@@ -861,6 +861,58 @@ int __syscall_rmdir(intptr_t path) {
   return __syscall_unlinkat(AT_FDCWD, path, AT_REMOVEDIR);
 }
 
+// wasmfs_unmount is similar to __syscall_unlinkat, but assumes AT_REMOVEDIR is true and will unlink nonempty directories.
+int wasmfs_unmount(int dirfd, intptr_t path) {
+  std::string_view p((char*)path);
+  // Ignore trailing '/'.
+  while (!p.empty() && p.back() == '/') {
+    p.remove_suffix(1);
+  }
+  if (p.size() >= 2 && p.substr(p.size() - 2) == std::string_view("/.")) {
+    return -EINVAL;
+  }
+
+  auto parsed = path::parseParent((char*)path, dirfd);
+  if (auto err = parsed.getError()) {
+    return err;
+  }
+  auto& [parent, childNameView] = parsed.getParentChild();
+  std::string childName(childNameView);
+  auto lockedParent = parent->locked();
+  auto file = lockedParent.getChild(childName);
+  if (!file) {
+    return -ENOENT;
+  }
+  // Disallow removing the root directory, even if it is empty.
+  if (file == wasmFS.getRootDirectory()) {
+    return -EBUSY;
+  }
+
+  auto lockedFile = file->locked();
+  if (auto dir = file->dynCast<Directory>()) {
+    if (parent->getBackend() == dir->getBackend()) {
+      // The child is not a valid mountpoint.
+      return -EINVAL;
+    }
+  } else {
+    // A normal file or symlink.
+    return -ENOTDIR;
+  }
+
+  // Cannot unlink/rmdir if the parent dir doesn't have write permissions.
+  if (!(lockedParent.getMode() & WASMFS_PERM_WRITE)) {
+    return -EACCES;
+  }
```

Comment:
Is there an existing test for this? If not, I think we can remove it. The man page for `umount` does not mention file permissions at all and unmounting a file system does not actually modify the underlying directory data, so it makes sense that permissions wouldn't need to be involved.

Reply:
There wasn't a test since I copied this section directly from `__syscall_unlinkat`, so I agree that it's not necessary.